### PR TITLE
Make equality checks faster and fix hash

### PR DESCRIPTION
**Pull request by oxinabox**
solves #15 

With this PR:
```julia
julia> @btime $(ShortString15("abc")) == "abc";
  10.935 ns (0 allocations: 0 bytes)

julia> @btime $(ShortString15("abc")) == "abd";
  11.152 ns (0 allocations: 0 bytes)
```

Without this PR
```julia
julia> @btime $(ShortString15("abc")) == "abc";
  79.701 ns (3 allocations: 224 bytes)

julia> @btime $(ShortString15("abc")) == "abd";
  80.259 ns (3 allocations: 224 bytes)
```

This also correct the `hash` closing #21

### CODE REVIEW

File: src/base.jl [hunk 1]
```diff
@@ -35,7 +35,7 @@ Base.isvalid(s::ShortString, i::Integer) = isvalid(String(s), i)
 Base.iterate(s::ShortString) = iterate(String(s))
 Base.iterate(s::ShortString, i::Integer) = iterate(String(s), i)
 Base.lastindex(s::ShortString) = sizeof(s)
-Base.ncodeunits(s::ShortString) = ncodeunits(String(s))
+Base.ncodeunits(s::ShortString) = sizeof(s)
```
Comment (oxinabox) [inline src/base.jl:38]: Is this right?
If so:
I am not sure we should be overloading `sizeof` or just `ncodeunits`

**Comment by oxinabox** [comment]
once we have this, it would be useful to tag a release as I want to use this for something

File: src/base.jl (hunk 1 above)
Comment (xiaodaigh) [reply src/base.jl:38]: I stand corrected. `sizeof` and `ncodeunits` should be the same here. I was confused. I thought `ncodeunits` counted characters but `length` does

File: src/base.jl [hunk 2]
```diff
@@ -35,18 +35,19 @@ Base.isvalid(s::ShortString, i::Integer) = isvalid(String(s), i)
...
 Base.print(s::ShortString) = print(String(s))
 Base.show(io::IO, str::ShortString) = show(io, String(str))
-Base.sizeof(s::ShortString{T}) where T = Int(s.size_content & size_mask(T))
+Base.sizeof(s::ShortString{T}) where T = Int(s.size_content & (size_mask(s) % UInt))
```
Comment (xiaodaigh) [inline src/base.jl:41]: Please help me understand why is `%` necessary? I must be missing something obvious.

File: src/base.jl [hunk 3]
```diff
@@ -59,9 +60,27 @@ size_mask(T) = UInt(exp2(4*size_nibbles(T)) - 1)
...
 
-==(s::ShortString, b::AbstractString) = begin
-    String(s) == b
+function ==(s::ShortString{S}, b::Union{String, SubString{String}}) where S
+    ncodeunits(b) == ncodeunits(s) || return false
+    return s == ShortString{S}(b)
 end
+function ==(s::ShortString, b::AbstractString)
+    # Could be a string type that might not use UTF8 encoding and that we don't have a
+    # constructor for. Defer to equality that that type probably has defined on `String`
```
Comment (xiaodaigh) [inline src/base.jl:69]:
```suggestion
    # constructor for. Defer to equality that type probably has defined on `String`
```

File: src/base.jl [hunk 4]
```diff
@@ -59,9 +60,27 @@ size_mask(T) = UInt(exp2(4*size_nibbles(T)) - 1)
...
+    return String(s) == b
+end
+
+==(a::AbstractString, b::ShortString) = b == a
+function ==(a::ShortString{S}, b::ShortString{S}) where S
+    return a.size_content == b.size_content
+end
+function ==(a::ShortString{A}, b::ShortString{B}) where {A,B}
+    ncodeunits(a) == ncodeunits(b) || return false
+    # compare if equal after droppign size bits and
```
Comment (xiaodaigh) [inline src/base.jl:79]:
```suggestion
    # compare if equal after dropping size bits and
```

File: src/hash.jl [hunk 5]
```diff
@@ -2,6 +2,4 @@ export hash
 
 import Base.hash
 
-Base.hash(x::ShortString, args...; kwargs...) = hash(x.size_content, args...; kwargs...)
-
-Base.hash(x::ShortString, h::UInt) = hash(x.size_content, h)
+Base.hash(x::ShortString, h::UInt) = hash(String(x), h)
```
[the remark this answers is not in the document]
Comment (xiaodaigh) [reply src/hash.jl:5]: I think it's ok for now.

**Review by xiaodaigh — APPROVED** [review]

File: src/base.jl (hunk 2 above)
Comment (xiaodaigh) [reply src/base.jl:41]: I got it.

**Comment by xiaodaigh** [comment]
Thanks!

File: src/hash.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (oxinabox) [reply src/hash.jl:5]: I think we have nerdsniped 
@ScottPJones into making it do something awesomely fast with MurmurHash3.jl
Looking forward to that PR

**Review by oxinabox — COMMENTED** [review]
..

File: src/hash.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaodaigh) [reply src/hash.jl:5]: very nice!

**Pull request merged**